### PR TITLE
Ins 731 objects routing

### PR DESCRIPTION
**Pull request by egorikas**
**- What I did**
Adding token to the SignedMessage
Moving Target/Role to the SignedMessageHeader
Checking token in the message bus

**- How I did it**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: core/message/messagehelpers.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	case *SignedMessage:
+		return ExtractTarget(t.Msg)
+	default:
+		panic(fmt.Sprintf("unknow message type - %v", t))
+	}
+}
+
+func ExtractRole(msg core.Message) core.JetRole {
+	switch t := msg.(type) {
+	case *BootstrapRequest:
```
[the remark this answers is not in the document]
Comment (egorikas) [reply core/message/messagehelpers.go]: solved

File: core/message/routingtoken.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	sign, err := crypto_helper.Sign(tokenBuffer.Bytes(), key)
+	if err != nil {
+		panic(err)
+	}
+	token.Sign = sign
+	return token
+}
+
+// ValidateToken checks that a routing token is valid
+func ValidateToken(pubKey *ecdsa.PublicKey, msg core.Parcel) error {
```
Comment (andreyromancev) [inline core/message/routingtoken.go]: msg core.Parcel - message or parcel?

Comment (andreyromancev) [reply core/message/routingtoken.go]: Maybe we should call it ValidateParcel?

File: core/messagebus.go [hunk 3]
```diff
@@ -49,15 +45,25 @@ type Signature interface {
...
 	Message() Message
 	Context(context.Context) context.Context
 	// Pulse returns pulse when message was sent.
 	Pulse() PulseNumber
+	// GetToken returns a routing-token of the message
+	GetToken() MessageToken
+}
+
+// MessageToken is the base interface for the routing token
+type MessageToken interface {
```
Comment (andreyromancev) [inline core/messagebus.go]: RoutingToken

File: core/message/routingtoken.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	serialized, err := ToBytes(msg.Message())
+	if err != nil {
+		return errors.Wrap(err, "filed to serialize message")
+	}
+	msgHash := hash.SHA3Bytes256(serialized)
+	token := RoutingToken{
```
Comment (andreyromancev) [inline core/message/routingtoken.go]: Maybe we should write hash from fields, not from token struct. That way we won't have to create struct to check.

File: ledger/artifactmanager/handler.go [hunk 5]
```diff
@@ -31,7 +31,7 @@ import (
 	"github.com/insolar/insolar/ledger/storage"
 )
 
-type internalHandler func(ctx context.Context, pulseNumber core.PulseNumber, genericMsg core.SignedMessage) (core.Reply, error)
+type internalHandler func(ctx context.Context, pulseNumber core.PulseNumber, genericMsg core.Parcel) (core.Reply, error)
```
Comment (andreyromancev) [inline ledger/artifactmanager/handler.go]: genericMsg -> parcel

**Pull request merged**